vowpalwabbit/v_array.h

```diff
@@ -36,76 +36,157 @@ struct v_array
+  // instead of indices. But these fn signatures follow our usage better.
+  // These functions do not check bounds, undefined behavior if they are called
+  // on out-of-bounds indices
+  // insert before the indexed element
+  inline void insert(size_t idx, const T& elem)
   {
     if (_end == end_array)
-      resize(2 * (end_array - _begin) + 3);
+      reserve_nocheck(2 * capacity() + 3);
     _end++;
+    memmove(&_begin[idx + 1], &_begin[idx], (size() - (idx + 1)) * sizeof(T));
+    _begin[idx] = elem;
   }
-  T& operator[](size_t i) const { return _begin[i]; }
+  // erase indexed element
+  inline void erase(size_t idx)
+  {
+    _begin[idx].~T();
+    memmove(&_begin[idx], &_begin[idx + 1], (size() - (idx + 1)) * sizeof(T));
+    --_end;
+  }
+
+  T& operator[](size_t i) { return _begin[i]; }
+  const T& operator[](size_t i) const { return _begin[i]; }
   inline size_t size() const { return _end - _begin; }
+  inline size_t capacity() const { return end_array - _begin; }
+
+  // change the number of elements in the vector
   void resize(size_t length)
   {
-    if ((size_t)(end_array - _begin) != length)
+    auto old_size = size();
+    // if new length is smaller than current size destroy the excess elements
+    for (auto idx = length; idx < old_size; ++idx)
     {
-      size_t old_len = _end - _begin;
-      T* temp = (T*)realloc(_begin, sizeof(T) * length);
-      if ((temp == nullptr) && ((sizeof(T) * length) > 0))
+      _begin[idx].~T();
+    }
+    if (capacity() < length)
+    {
+      reserve(length);
+    }
+    _end = _begin + length;
+    // default construct any newly added elements
+    // TODO: handle non-default constructable objects
+    // requires second interface
+    /*
+    for (auto idx = old_size; idx < length; ++idx)
+    {
+      new (&_begin[idx]) T();
+    }
+    */
+  }
+
+  void shrink_to_fit()
+  {
+    if (size() < capacity())
+    {
+      if (empty())
       {
-        THROW_OR_RETURN("realloc of " << length << " failed in resize().  out of memory?");
+        // realloc on size 0 doesn't have a specified behavior
+        // just shrink to 1 for now (alternatively, call delete_v())
+        reserve_nocheck(1);
       }
       else
-        _begin = temp;
-      if (old_len < length && _begin + old_len != nullptr)
-        memset(_begin + old_len, 0, (length - old_len) * sizeof(T));
-      _end = _begin + old_len;
-      end_array = _begin + length;
+      {
+        reserve_nocheck(size());
+      }
     }
   }
+  // reserve enough space for the specified number of elements
+  void reserve(size_t length)
+  {
+    if (capacity() < length)
+      reserve_nocheck(length);
+  }
 
   void clear()
   {
     if (++erase_count & erase_point)
     {
-      resize(_end - _begin);
+      shrink_to_fit();
       erase_count = 0;
     }
     for (T* item = _begin; item != _end; ++item) item->~T();
     _end = _begin;
   }
+  void reset()
+  {
```

Comment:
This will leak if we don't clear what's inside

Reply:
Correct. There are a bunch of places we want to reset the pointers without deleting the buffer

Reply:
Ah this is so dangerous... I am pretty sure this will go away once we can move right?